x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -619,6 +680,32 @@ Role getRole() {
         public RBACAuthorizationInfo getAuthenticatedUserAuthorizationInfo() {
             return authenticatedUserAuthorizationInfo;
         }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            RBACAuthorizationInfo that = (RBACAuthorizationInfo) o;
+            if (this.role.equals(that.role) == false) {
+                return false;
+            }
```

Comment:
The identity comparison does not work well for API keys in some cases because `LimitedRole` is created on the fly

https://github.com/elastic/elasticsearch/blob/83125377bd8ee1ac96aed1c1c43c74d5b13e6cde/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java#L301-L302

I don't have a great suggestion. Maybe let the `LimitedRole` keeps a reference of the keyRole and have a `equals` method to compare the identities of the underlying keyRole and ownerRole?

Reply:
It's not an identity comparison - it's `.equals` with a negation (`== false`)

Reply:
But `Role` inherits `equals` from `Object` which only performs an identity check?

Reply:
Thanks for the clarification. I've added equals methods to `Role` and `LimitedRole`. They're not perfect (comparing permissions is hard) but they should do the right thing for API Keys.